doc/freenukum.6
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.\"

.TH freenukum 6

.SH NAME
freenukum \- Jump'n Run game

.SH SYNOPSIS
.B freenukum

.SH DESCRIPTION
The \fBFreenukum\fR game is a free software implementation of the
Duke Nukum 1 game. The player must go through the levels, shoot
enemy robots and find keys and access cards in order to unlock
barriers and reach the exit door.

.SH DATA FILES
The original level files
of the Duke Nukum 1 game need to be copied to
\fB~/.freenukum/data\fR. This can either be the shareware episode
only which is available for download free of charge from
http://www.3drealms.com/duke1/index.html or the
full version which contains three episodes and can also be
acquired from this website. In order to unpack the data
from the original game, a DOS environment is needed.
This can either be \fBdosbox\fR, \fBfreedos\fR or
an installation of \fBDOS\fR or \fBWindows\fR.

.SH ITEMS
The hero can also collect several different items, which are
listed below. The collected items are shown in the right
section of the screen.

The more \fBfirepower\fR the hero has,
the more shots can stay inside the window simultaneously.
This means the hero can shoot faster and doesn not have
to wait until a shot leaves the visible area.

\fBKeys\fR exist in several different colors. Every key
fits into the keyhole of the same color as the key is.
The activation of the keyhole unlocks the door which
is linked to it.

The \fBAccess Card\fR work just the same way as the keys.
Once used on the access card slot, it opens the
laser beam which keeps the hero from passing on.

The \fBBoot\fR allows the hero to jump higher and
reach areas he wouldn't be able to access otherwise.

The \fBGlove\fR allows the hero to activate a special switch
which enables a floor where there would be none. This way
the hero can walk over the floor in order to finish his
goals.

The \fBClamps\fR allow the hero to attach to some special ceilings.

.SH CONTROL
Freenukum is controlled by the keyboard. The \fBleft arrow\fR
and the \fBright arrow\fR are used for walking left and
right. The \fBleft CTRL key\fR is used for jumping, and the
\fBleft ALT key\fR is used for shooting. In order to
activate something (keyhole, exit door, elevator etc.) the
\fBup arrow\fR is used.

You can switch the game to fullscreen by pressing the
\fBf key\fR.

.SH OPTIONS
No options are currently available to freenukum.

.SH AUTHOR
Written by Wolfgang Silbermayr.
This manual page was originally written by Wolfgang Silbermayr <wolfgang@silbermayr.at>
.br
Visit official homepage: http://launchpad.net/freenukum/

.SH COPYRIGHT
Copyright \(co 2007-2008 The Freenukum Team.
.br
This is Free Software; this software is licensed under the GPL version 3, as published by the Free Software Foundation.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


ChangeLog
	- Implemented fire wheel bot (LP: #242134)
0.2.9
	- Implemented fan wheels (LP: #242128)
	- No more solid parts behind cameras (LP: #242133)
	- Added fullscreen mode (LP: #245274)
	- Implemented glove slot which can expand floor
	- Soda can flys away when shot by hero (LP: #242118)
	- Freenukum can now also load lowercase named data files (LP: #245269)
0.2.8
	- Fixing compile bug on AMD64 (LP: #244763)
0.2.7
	- Implemented shots by enemies (LP: #243990)
	- Implemented bombs (LP: #242116)
	- Removed dependency on glib (LP: #243956)
	- Added background (LP: #240832)
	- Added particle fireworks (LP: #242117)
	- Added menu shortcut .desktop file
	- Added tankbot (LP: #241105)
0.2.6
	- Using BitsPerPixel and SDL-Flags from initial screen (LP: #243352)
	- Calling SDL_Quit on exit of the program (LP: #243597)
	- Distribution tarball includes manpage (LP: #243296)
	- fn_test_* programs can now be compiled using
	  ./configure --enable-testprograms
0.2.5
	- Made freenukum compile on AMD64 (LP: #242228)
0.2.4
	- Implemented flamethrower
	- Shots can no longer go through solid parts
	- Added ACME stone
	- Added Manpage
	- Implemented mines
	- Implemented floor which crashes when hero steps upon it twice
0.2.3
	- Implemented balloons
	- Implemented wall-crawling green bots
	- When hero gets hurt he has an immunity phase before he is hurt again
	- Made it possible to change to other episodes (if installed).
0.2.2
	- Implemented simple robot
	- Made hero motion model more similar to original game.
	- Elevators implemented
0.2.1
	- Hero can fetch guns for more firepower.
	- Hero can fetch boots and then jump higher.
	- Hero can use teleporters.
	- Hero can shoot boxes.
	- Keyhole stops blinking after key insert.
	- Hero can walk through opened doors immediately instead of having
	  to wait until the animation finishes.
	- Hero earns score for fetching keys.
	- Keys can now be used.
	- After finishing a level, we go on to the next instead of returning
	  to the title screen

0.2
	- Exit doors work now

0.1
	- Initial development release


src/freenukum.c
/*******************************************************************
 *
 * Project: FreeNukum 2D Jump'n Run
 * File:    Main game starting file
 *
 * *****************************************************************
 *
 * Copyright 2007-2008 Wolfgang Silbermayr
 *
 * *****************************************************************
 *
 * This file is part of Freenukum.
 * 
 * Freenukum is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Freenukum is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *******************************************************************/

#include <SDL.h>
#include <stdlib.h>
#include <sys/types.h>
#include <dirent.h>
#include <errno.h>
#include <string.h>
#include <sys/stat.h>

/* --------------------------------------------------------------- */

#include "fn.h"
#include "fn_msgbox.h"
#include "fn_tilecache.h"
#include "fn_error_cmdline.h"
#include "fn_error.h"
#include "fn_picture_splash.h"
#include "fn_mainmenu.h"
#include "fn_infobox.h"
#include "fn_settings.h"
#include "fn_game.h"
#include "fn_data.h"

/* --------------------------------------------------------------- */

int main(int argc, char ** argv)
{
  SDL_Surface * screen; /* the main screen surface */

  long int pixelsize; /* pixel size */
  Uint8 fullscreen; /* Play fullscreen? */

  fn_tilecache_t tilecache; /* the main tilecache */

  int res; /* results are stored here */

  int retval = 1; /* The return value from the program. */

  int choice = 0; /* choice of the main menu */

  char * homepath; /* home directory of the user */
  char configpath[1024]; /* path to the settings directory */
  char datapath[1024]; /* path to the original game data */
  char configfilepath[1024]; /* path to the config file */

  DIR * configdir; /* directory pointer to the config directory */
  DIR * datadir; /* directory pointer to the data directory */

  fn_settings_t * settings = NULL; /* the settings struct */

  Uint8 episode = 1; /* The episode currently being played. */

  char backgroundfile[10] = "DN.DN1";

  Uint32 videoflags = FN_SURFACE_FLAGS;

/* --------------------------------------------------------------- */

  fn_error_set_handler(fn_error_print_commandline);

  /* retrieve home directory path */
  homepath = getenv("HOME");
  if (homepath == NULL) {
    fn_error_print("$HOME environment variable is not set.");
    goto homepath_failed;
  }

  snprintf(configpath, 1024, "%s%s", homepath, "/.freenukum");
  snprintf(datapath, 1024, "%s%s", configpath, "/data/");
  snprintf(configfilepath, 1024, "%s%s", configpath, "/config");

  /* check if .freenukum directory exists and create it if
   * it does not exist yet */
  configdir = opendir(configpath);
  if (configdir == NULL) {
    res = mkdir(configpath, S_IRUSR | S_IWUSR | S_IXUSR
        | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH);
    if (res) {
      fn_error_printf(1024, "Could not create the directory %s: %s",
          configpath, strerror(errno));
      goto configdir_failed;
    }
  }

  /* check if data directory exists and create it if
   * it does not exist yet */
  datadir = opendir(datapath);
  if (datadir == NULL) {
    res = mkdir(datapath, S_IRUSR | S_IWUSR | S_IXUSR
        | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH);
    if (res) {
      fn_error_printf(1024, "Could not create the directory %s: %s",
          datapath, strerror(errno));
      goto datadir_failed;
    }
  }

  /* load the settings from the config file and
   * create the file if it does not exist */
  settings = fn_settings_new_from_file(configfilepath);
  if (settings == NULL) {
    settings = fn_settings_new();
    res = fn_settings_store(settings, configfilepath);
    if (!res) {
      fn_error_printf(1024, "Could not create the settings file %s: %s",
          configfilepath, strerror(errno));
      goto settings_failed;
    }
  }

  /* load pixelsize from config */
  fn_settings_get_longint_with_default(settings, "pixelsize",
      &pixelsize, FN_DEFAULT_PIXELSIZE);
  fn_settings_get_bool_with_default(settings, "fullscreen",
      &fullscreen, FN_DEFAULT_FULLSCREEN);

  if (fullscreen) {
    videoflags |= SDL_FULLSCREEN;
  }

  /* initialize SDL */
  if (SDL_Init(SDL_INIT_VIDEO | SDL_INIT_TIMER) == -1) {
    fn_error_printf(1024, "Could not initialize SDL: %s", SDL_GetError());
    goto sdl_init_failed;
  }

  /* initialize SDL window */
  screen = SDL_SetVideoMode(
      FN_WINDOW_WIDTH * pixelsize,
      FN_WINDOW_HEIGHT * pixelsize,
      FN_COLOR_DEPTH,
      videoflags);
  if (screen == NULL) {
    fn_error_printf(1024, "Can't set video mode: %s", SDL_GetError());
    goto sdl_videomode_failed;
  }

  /* set window caption */
  SDL_WM_SetCaption("Freenukum " VERSION, "Freenukum " VERSION);

  /* check if all data is present */
  int episodes = 0;
  while (res != -1) {
    res = fn_data_check(datapath, episodes + 1);
    if (res == -1 && episodes == 0) {
      /* we found zero episodes */
      printf("Could not load any episode.\n");
      goto data_check_failed;
    } else if (res != -1) {
      /* we found the currently requested episode */
      episodes++;
    }
  }
  printf("Found %d episodes.\n", episodes);

  /* initialize the tile cache */
  fn_tilecache_init(&tilecache, (Uint8)pixelsize);
  res = fn_tilecache_loadtiles(
      &tilecache,
      screen->flags,
      screen->format,
      datapath);
  if (res == -1) {
    fn_error_printf(1024, "Could not load tiles.\n"
        "Copy the original game files to %s.\n"
        "They can be downloaded free of charge from\n"
        "http://www.3drealms.com/duke1/index.html" , datapath);
    /* TODO show text in screen maybe? */
    goto tilecache_failed;
  }


  /* show the splash screen */
  res = fn_picture_splash_show(
      datapath,
      backgroundfile,
      (Uint8)pixelsize,
      screen);
  if (!res) {
    fn_error_printf(1024, "Could not show splash screen.\n");
    goto splashscreen_failed;
  }

  /* show the main menu */
  while (choice != FN_MENUCHOICE_QUIT) {
    choice = fn_mainmenu(&tilecache, (Uint8)pixelsize, screen);
    switch(choice) {
      case FN_MENUCHOICE_START:
        fn_game_start(
            (Uint8)pixelsize, &tilecache, screen, datapath, episode,
            settings);
        res = fn_picture_splash_show(datapath,
            backgroundfile,
            (Uint8)pixelsize,
            screen);
        break;
      case FN_MENUCHOICE_RESTORE:
        fn_infobox_show((Uint8)pixelsize, &tilecache, screen,
            "Restore not implemented yet.\n");
        break;
      case FN_MENUCHOICE_INSTRUCTIONS:
        fn_infobox_show((Uint8)pixelsize, &tilecache, screen,
            "Instructions not implemented yet.\n");
        break;
      case FN_MENUCHOICE_ORDERINGINFO:
        fn_infobox_show((Uint8)pixelsize, &tilecache, screen,
            "Orderinginfo not implemented yet.\n");
        break;
      case FN_MENUCHOICE_SETUP:
        fn_infobox_show((Uint8)pixelsize, &tilecache, screen,
            "Setup not implemented yet.\n");
        break;
      case FN_MENUCHOICE_FULLSCREENTOGGLE:
        {
          int res = SDL_WM_ToggleFullScreen(screen);
          if (res) {
            fullscreen = (fullscreen + 1) % 2;
            fn_settings_set_bool(settings, "fullscreen", fullscreen);
          }
        }
        break;
      case FN_MENUCHOICE_EPISODECHANGE:
        {
          episode++;
          if (episode > 3) {
            episode = 1;
          }
          snprintf(backgroundfile,
              7, "DN.DN%d", episode);
          res = fn_picture_splash_show(datapath,
              backgroundfile, (Uint8)pixelsize, screen);
          if (res == 0) {
            fn_infobox_show((Uint8)pixelsize, &tilecache, screen,
                "You don't have this episode installed.\n"
                "We stay in episode 1\n");
            episode = 1;
            snprintf(backgroundfile,
                7, "DN.DN%d", episode);
            res = fn_picture_splash_show(datapath,
                backgroundfile, (Uint8)pixelsize, screen);
          }
        }
        break;
      case FN_MENUCHOICE_HIGHSCORES:
        fn_infobox_show((Uint8)pixelsize, &tilecache, screen,
            "Highscores not implemented yet.\n");
        break;
      case FN_MENUCHOICE_PREVIEWS:
        fn_infobox_show((Uint8)pixelsize, &tilecache, screen,
            "Previews not implemented yet.\n");
        break;
      case FN_MENUCHOICE_VIEWUSERDEMO:
        fn_infobox_show((Uint8)pixelsize, &tilecache, screen,
            "Userdemo not implemented yet.\n");
        break;
      case FN_MENUCHOICE_TITLESCREEN:
        res = fn_picture_splash_show(datapath,
            backgroundfile,
            (Uint8)pixelsize,
            screen);
        break;
      case FN_MENUCHOICE_CREDITS:
        fn_infobox_show((Uint8)pixelsize, &tilecache, screen,
            "Credits not implemented yet.\n");
        break;
      default:
        break;
    }
  }

  res = fn_settings_store(settings, configfilepath);
  if (!res) {
    fn_error_printf(1024, "Could not store settings to %s: %s",
        configfilepath, strerror(errno));
  }

  retval = 0;

splashscreen_failed:
  fn_tilecache_destroy(&tilecache);
tilecache_failed:
data_check_failed:
sdl_videomode_failed:
  SDL_Quit();
sdl_init_failed:
  fn_settings_free(settings);
settings_failed:
  closedir(datadir);
datadir_failed:
  closedir(configdir);
configdir_failed:
homepath_failed:
  return retval;
}
